modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+
+    private final Settings settings;
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ResultDeduplicator<TransportRequest, Void> transportActionsDeduplicator;
+    private final LongSupplier nowSupplier;
+    private final Clock clock;
+    private volatile boolean isMaster = false;
+    private volatile TimeValue pollInterval;
+    private SchedulerEngine.Job scheduledJob;
+    private final SetOnce<SchedulerEngine> scheduler = new SetOnce<>();
+    // we use this rollover supplier to facilitate testing until we'll be able to read the
+    // rollover configuration from a cluster setting
+    private Function<String, RolloverRequest> defaultRolloverRequestSupplier;
+
+    public DataLifecycleService(
+        Settings settings,
+        Client client,
+        ClusterService clusterService,
+        Clock clock,
+        ThreadPool threadPool,
+        LongSupplier nowSupplier
+    ) {
+        this.settings = settings;
+        this.client = client;
+        this.clusterService = clusterService;
+        this.clock = clock;
+        this.transportActionsDeduplicator = new ResultDeduplicator<>(threadPool.getThreadContext());
+        this.nowSupplier = nowSupplier;
+        this.scheduledJob = null;
+        this.pollInterval = DLM_POLL_INTERVAL_SETTING.get(settings);
+        this.defaultRolloverRequestSupplier = this::getDefaultRolloverRequest;
+    }
+
+    /**
+     * Initializer method to avoid the publication of a self reference in the constructor.
+     */
+    public void init() {
+        clusterService.addListener(this);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(DLM_POLL_INTERVAL_SETTING, this::updatePollInterval);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        final boolean prevIsMaster = this.isMaster;
+        if (prevIsMaster != event.localNodeMaster()) {
+            this.isMaster = event.localNodeMaster();
+            if (this.isMaster) {
+                // we weren't the master, and now we are
+                onBecomeMaster();
+            } else {
+                // we were the master, and now we aren't
+                cancelJob();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        SchedulerEngine engine = scheduler.get();
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(DATA_LIFECYCLE_JOB_NAME)) {
+            logger.trace("DLM job triggered: " + event.getJobName() + ", " + event.getScheduledTime() + ", " + event.getTriggeredTime());
+            run(clusterService.state());
+        }
```

Comment:
Should we add a `if (this.isMaster)` check here, to ensure that we don't ever have an issue (through some other bug) where a scheduler is running a DLM task on a non-master node?